### PR TITLE
Adds docker.latest_running state

### PR DESCRIPTION
**Pull request by jfindlay**
Ensures the container is running with the correct configuration and its
image is the latest available.  Rebase of #19837.

### CODE REVIEW

File: salt/states/dockerio.py [hunk 1]
```diff
@@ -108,13 +108,17 @@
...
+import inspect
 import logging
 
 # Import salt libs
+from salt.exceptions import CommandExecutionError, SaltInvocationError
 from salt.ext.six import string_types
 import salt.utils
 import salt.ext.six as six
 
+import docker
```
Comment (cachedout) [inline salt/states/dockerio.py]: This will cause state runs to stacktrace if the docker module is not installed.

Comment (thatch45) [inline salt/states/dockerio.py]: This should just put a stack trace in the log, but not crash the state run, if it crashes the state run that needs to be fixed.
but yes, this needs to be gated

Comment (cachedout) [inline salt/states/dockerio.py]: Yes, in the log. It does not affect the state run itself.

**Comment by ticosax** [comment]
What is the reasoning behind this state when it could be achieved with requisites:

```sls

my-service-image:
  docker.pulled:
    - name: registry/my-service:latest
    - force: true

my-service-container:
  docker.installed:
    - image: registry/my-service:latest
    - watch:
        - docker: my-service-image

my-service:
  docker.running:
    - container: aviary-backend
    - watch:
        - docker: my-service-container
```
here the watchers logic will create a new container based on updated image only if a new one has been downloaded, stop and destroy the old container and start the new container.

**Comment by jfindlay** [comment]
@ticosax, I am not very familiar with the docker code.  You may want to ask @ipmb.

**Comment by ipmb** [comment]
@ticosax I was a little surprised this was merged tbh, I was hoping to start some discussion on it here: https://github.com/saltstack/salt/pull/19837

As to what's different, it bundles up a common pattern into a single state. It also checks the running config against the requested config for changes and restarts the running container if necessary.

**Comment by ticosax** [comment]
I understand having a single state might be conveniant to solve this use-case. But looking at the diff you introduce, I'm in favor of keeping the code simpler and reverting this pull request at the expense of documenting how we solve this use-case with in more ``salty`` way.
Basically, we could take the example I provided and put it in the documentation.

**Comment by jfindlay** [comment]
@ticosax, If all of the functionality added by this pull request can be achieved by states, please feel free to submit a pull request to this effect.  What do you think, @ipmb?  Can [this](https://github.com/saltstack/salt/pull/19837#issuecomment-71087491) also be achieved in a less complex manner?

**Comment by ipmb** [comment]
I'm fine with whatever. I do think there's a couple things of value here that weren't there before:

* [Checking if the configuration has changed](https://github.com/saltstack/salt/blob/94995c6b9fedf18ba26a0471d255ad715f9c1e47/salt/states/dockerio.py#L194-L236). This is functionality that doesn't currently exist and leads to unexpected behavior imo. If you change the configuration a running container today, Salt just ignores it.
* [Accepting port definitions in the same format as the Docker CLI](https://github.com/saltstack/salt/blob/94995c6b9fedf18ba26a0471d255ad715f9c1e47/salt/states/dockerio.py#L294-L325). No new functionality, but it makes the states much easier to use in my opinion. I'm guessing users will have a better familiarity with the Docker CLI than the Python client.
* [Accepting volume definitions in the same format as the Docker CLI](https://github.com/saltstack/salt/blob/94995c6b9fedf18ba26a0471d255ad715f9c1e47/salt/states/dockerio.py#L269-L291). Same as the port definitions above.

**Comment by ticosax** [comment]
I do not think your approach is wrong, but I have this three concerns:

 - ` Checking if the configuration has changed` should be handled by each state like in `docker.pulled`, `docker.installed` or `docker.running`. If they does not behave as expected it means there is a bug and should they be fixed accordingly. Relying on `docker.latest_running` to achieve configuration change detection, is a workaround.

 - `regarding mocking CLI parameters`, you implemented a new parser to get your arguments when yaml can already do it. Those new parser will now have to be maintained and tested and so on.
The question is, does it worth it ? As a contributor of this salt module, I refer myself to docker-py, not the command line. It means I'm not interested in CLI style parameters.

- There is already a bug report for `docker.latest_running` #21580 , it just shows that changes must be motivated by solving a real life problem. I do not see yet what `docker.latest_running` is solving on its own.

**Comment by ipmb** [comment]
> Checking if the configuration has changed should be handled by each state like in docker.pulled, docker.installed or docker.running. If they does not behave as expected it means there is a bug and should they be fixed accordingly. Relying on docker.latest_running to achieve configuration change detection, is a workaround.

Agreed.

> regarding mocking CLI parameters, you implemented a new parser to get your arguments when yaml can already do it. Those new parser will now have to be maintained and tested and so on.
The question is, does it worth it ? As a contributor of this salt module, I refer myself to docker-py, not the command line. It means I'm not interested in CLI style parameters.

Personally, I found the Python API confusing coming from the Docker CLI, but people will figure it out either way.

> There is already a bug report for docker.latest_running #21580 , it just shows that changes must be motivated by solving a real life problem. I do not see yet what docker.latest_running is solving on its own.

Feel free to remove it. It was not my intent to have it merged, but to start the discussion we are having now :)

**Comment by ticosax** [comment]
It would have been great to have this discussion earlier indeed :)
You will realize that at saltstack they pretty merge everything, the good news is that the community react very quickly and fix bugs sometimes within the hour.

So, I think I will revert it, and document how to solve this use case the salty way. I own you that at least :)

thank you for the follow up.

**Pull request merged**